keras_nlp/models/xlnet/xlnet_tokenizer.py

```diff
@@ -0,0 +1,205 @@
+        self.eos_token_id = self.token_to_id(self.eos_token)
+        self.unk_token_id = self.token_to_id(self.unk_token)
+
+    def preprocess_text(self, inputs):
+        """Preprocesses the text. This method removes spaces and accents from the text."""
+
+        # remove space
+        outputs = " ".join(inputs.strip().split())
+        outputs = outputs.replace("``", '"').replace("''", '"')
+
+        # remove accents
+        outputs = unicodedata.normalize("NFKD", outputs)
+        outputs = "".join([c for c in outputs if not unicodedata.combining(c)])
+
+        return outputs
+
+    def tokenize(self, text):
+        """Tokenize a string."""
+
+        # check if there are multiple examples present or not
+        is_batched = isinstance(text, list)
+        if not is_batched:
+            text = [text]
+
+        tokenized_text = []
+        for each_text in text:
+            each_text = self.preprocess_text(each_text)
+            pieces = [
+                self.id_to_token(token_id)
+                for token_id in super().tokenize(each_text)
+            ]
+
+            new_pieces = []
+            for piece in pieces:
+                if (
+                    len(piece) > 1
+                    and piece[-1] == str(",")
+                    and piece[-2].isdigit()
+                ):
+                    cur_pieces = [
+                        self.id_to_token(cur_piece_id)
+                        for cur_piece_id in super().tokenize(
+                            piece[:-1].replace("▁", "")
+                        )
+                    ]
+                    if piece[0] != "▁" and cur_pieces[0][0] == "▁":
+                        if len(cur_pieces[0]) == 1:
+                            cur_pieces = cur_pieces[1:]
+                        else:
+                            cur_pieces[0] = cur_pieces[0][1:]
+                    cur_pieces.append(piece[-1])
+                    new_pieces.extend(cur_pieces)
+                else:
+                    new_pieces.append(piece)
+
+            new_pieces = [
+                self.token_to_id(new_piece_token)
+                for new_piece_token in new_pieces
+            ]
+            # add sep_token and cls_token in the end.
+            new_pieces.extend([self.sep_token_id, self.cls_token_id])
+
+            tokenized_text.append(new_pieces)
+
+        # if there are multiple examples present, then return a `tf.RaggedTensor`.
+        if is_batched:
+            return tf.ragged.constant(tokenized_text)
+
+        return tokenized_text[0]
+
+    def detokenize(self, inputs):
+        """Detokenize the input_ids into text."""
+
+        outputs = super().detokenize(inputs)
+
+        outputs = tf.strings.regex_replace(outputs, self.cls_token, "")
+        outputs = tf.strings.regex_replace(outputs, self.sep_token, "")
+        outputs = tf.strings.regex_replace(outputs, self.mask_token, "")
+        outputs = tf.strings.regex_replace(outputs, self.pad_token, "")
```

Comment:
There is one difference in `detokenize` where contrary to the `HuggingFace` implementation, `keras_nlp` tokenizer removes the `<cls>` or `<sep>` tokens at the end,

For an example - 
```python
from transformers import XLNetTokenizer

tokenizer_hf = XLNetTokenizer.from_pretrained("xlnet-base-cased")
text = "the quick brown fox"
print(tokenizer_hf.decode(tokenizer_hf(text)["input_ids"]))
```
this will give us output - 
```
the quick brown fox<sep><cls>
```

the `keras-nlp` implementation will remove those <sep> and <cls> tokens and give us `the quick brown fox`.

Please let me know if I should change this design to strictly follow the HF or not.